datafusion/core/src/datasource/file_format/write.rs

```diff
@@ -315,58 +300,237 @@ pub(crate) async fn create_writer(
+        while let Some(maybe_batch) = data_stream.next().await {
+            match serializer.duplicate() {
+                Ok(mut serializer_clone) => {
+                    let handle = tokio::spawn(async move {
+                        let batch = maybe_batch?;
+                        let num_rows = batch.num_rows();
+                        let bytes = serializer_clone.serialize(batch).await?;
+                        Ok((num_rows, bytes))
+                    });
+                    tx.send(handle).await.map_err(|_| {
+                        DataFusionError::Internal(
+                            "Unknown error writing to object store".into(),
+                        )
+                    })?;
+                    if unbounded_input {
+                        tokio::task::yield_now().await;
+                    }
+                }
+                Err(_) => {
+                    return Err(DataFusionError::Internal(
+                        "Unknown error writing to object store".into(),
+                    ))
+                }
+            }
+        }
+        Ok(serializer)
+    });
+
+    let mut row_count = 0;
+    while let Some(handle) = rx.recv().await {
+        match handle.await {
+            Ok(Ok((cnt, bytes))) => {
+                match writer.write_all(&bytes).await {
+                    Ok(_) => (),
+                    Err(_) => {
+                        return Err((
+                            writer,
+                            DataFusionError::Internal(
+                                "Unknown error writing to object store".into(),
+                            ),
+                        ))
+                    }
+                };
+                row_count += cnt;
+            }
+            Ok(Err(e)) => {
+                // Return the writer along with the error
+                return Err((writer, e));
+            }
+            Err(_) => {
+                // Handle task panic or cancellation
+                return Err((
+                    writer,
+                    DataFusionError::Internal(
+                        "Serialization task panicked or was cancelled".into(),
+                    ),
+                ));
+            }
+        }
+    }
+
+    let serializer = match serialize_task.await {
+        Ok(Ok(serializer)) => serializer,
+        Ok(Err(e)) => return Err((writer, e)),
+        Err(_) => {
+            return Err((
+                writer,
+                DataFusionError::Internal("Unknown error writing to object store".into()),
+            ))
+        }
+    };
+    Ok((serializer, writer, row_count as u64))
+}
+
 /// Contains the common logic for serializing RecordBatches and
 /// writing the resulting bytes to an ObjectStore.
 /// Serialization is assumed to be stateless, i.e.
 /// each RecordBatch can be serialized without any
 /// dependency on the RecordBatches before or after.
 pub(crate) async fn stateless_serialize_and_write_files(
```

Comment:
The blocks inside the match `serialize_rb_stream_to_object_store` can be refactored into separate functions for better readability and maintainability.

```rust
pub(crate) async fn stateless_serialize_and_write_files(
    data: Vec<SendableRecordBatchStream>,
    mut serializers: Vec<Box<dyn BatchSerializer>>,
    mut writers: Vec<AbortableWrite<Box<dyn AsyncWrite + Send + Unpin>>>,
    single_file_output: bool,
    unbounded_input: bool,
) -> Result<u64> {
    if single_file_output {
        if serializers.len() != 1 || writers.len() != 1 {
            return internal_err!("single_file_output is true, but got more than 1 writer!");
        }
        return serialize_single_file(data, serializers.remove(0), writers.remove(0), unbounded_input).await;
    } else {
        if data.len() != writers.len() {
            return internal_err!("Did not get 1 writer for each output partition!");
        }
        return serialize_multiple_files(data, serializers, writers, unbounded_input).await;
    }
}
```